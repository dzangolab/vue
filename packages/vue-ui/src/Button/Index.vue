<template>
  <component
    :is="to ? 'a' : 'button'"
    :aria-label="ariaLabel ?? label"
    :class="buttonClassName"
    :disabled="!to && isDisabled"
    :href="to && !isDisabled ? to : null"
    :role="buttonRole"
    @click="!to && handleClick()"
  >
    <span v-if="iconLeft || !!slots.iconLeft" class="icon-left">
      <slot name="iconLeft">
        <i :class="iconLeft" />
      </slot>
    </span>

    <div v-if="children || label" class="label">
      <slot name="buttonContent">
        <span>{{ label }}</span>
      </slot>
    </div>

    <span v-if="iconRight || !!slots.iconRight" class="icon-right">
      <slot name="iconRight">
        <i :class="iconRight" />
      </slot>
    </span>

    <LoadingIcon v-if="loading" class="loading-button" />
  </component>
</template>

<script lang="ts">
export default {
  name: "ButtonElement",
};
</script>

<script setup lang="ts">
import { computed, useSlots } from "vue";

import LoadingIcon from "../components/LoadingIcon.vue";

const emits = defineEmits(["click"]);

const props = defineProps({
  ariaLabel: {
    default: "button",
    type: String,
  },
  children: {
    default: false,
    type: Boolean,
  },
  className: {
    default: "",
    type: String,
  },
  disabled: {
    default: false,
    type: Boolean,
  },
  iconLeft: {
    default: null,
    type: [String, Boolean],
  },
  iconRight: {
    default: null,
    type: [String, Boolean],
  },
  label: {
    default: null,
    type: String,
  },
  loading: {
    default: false,
    type: Boolean,
  },
  rounded: {
    default: false,
    type: Boolean,
  },
  severity: {
    default: "primary",
    type: String,
    validator: (value: string) =>
      [
        "primary",
        "secondary",
        "alternate",
        "success",
        "danger",
        "warning",
      ].includes(value),
  },
  size: {
    default: "medium",
    type: String,
    validator: (value: string) => ["small", "medium", "large"].includes(value),
  },
  to: {
    default: null,
    type: String,
  },
  title: {
    default: null,
    type: String,
  },
  variant: {
    default: "filled",
    type: String,
    validator: (value: string) =>
      ["outlined", "filled", "textOnly"].includes(value),
  },
});

const slots = useSlots();

const buttonClassName = computed(() => {
  return [
    "button",
    props.className,
    props.severity,
    props.size,
    props.variant,
    !(props.label || props.children) && "icon-only",
    props.loading && "loading",
    props.rounded && "rounded",
    isDisabled.value && "disabled",
  ]
    .filter(Boolean)
    .join(" ");
});

const buttonRole = computed(() => (props.to ? "link" : "button"));

const isDisabled = computed(() => props.disabled || props.loading);

function handleClick() {
  emits("click");
}
</script>

<style lang="css">
.button .loading-button {
  font-size: 4px;
  position: absolute;
  z-index: 999;
}
</style>
