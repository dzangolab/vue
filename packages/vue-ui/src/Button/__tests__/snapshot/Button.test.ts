import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import Button from "../../Index.vue";

describe("Button", () => {
  it("matches snapshot", () => {
    const wrapper = mount(Button, {
      props: {
        children: false,
        className: "rounded",
        iconLeft: true,
        iconRight: true,
        loading: false,
        label: "Submit",
        severity: "primary",
        size: "small",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
