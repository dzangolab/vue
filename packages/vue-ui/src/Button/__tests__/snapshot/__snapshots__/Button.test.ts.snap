// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<button
  aria-label="button"
  class="rounded"
  role="button"
>
  <span
    class="icon-left"
  >
    
    <i
      class=""
    />
    
  </span>
  <div
    class="label"
  >
    
    <span>
      Submit
    </span>
    
  </div>
  <span
    class="icon-right"
  >
    
    <i
      class=""
    />
    
  </span>
  <!--v-if-->
</button>
`;
